### PR TITLE
Reduce implementation code inside CScript

### PR DESCRIPTION
**Pull request by sipa**
This PR moves `FindAndDelete` and `GetOp2` out of CScript (the first is only used inside the interpreter and moved there, the second does not actually depend on any script specifics and works on any vector). Furthermore, all non-const-iterator versions of GetOp are replaced by const ones, removing a number of methods in the process.

The longer term goal here is making the script interpreter independent from the CScript representation.

Note for reviewers: both `FindAndDelete` and `GetScriptOp` are consensus critical.

### CODE REVIEW

File: src/script/script.h [hunk 1]
```diff
@@ -495,65 +497,14 @@ class CScript : public CScriptBase
...
     {
-        return GetOp2(pc, opcodeRet, &vchRet);
+        return GetScriptOp(pc, end(), opcodeRet, &vchRet);
     }
 
     bool GetOp(const_iterator& pc, opcodetype& opcodeRet) const
     {
-        return GetOp2(pc, opcodeRet, nullptr);
+        return GetScriptOp(pc, end(), opcodeRet, nullptr);
     }
```
Comment (Empact) [inline src/script/script.h]: Could also move the `GetOp` definitions and make `GetScriptOp` static.

File: src/script/script.cpp [hunk 2]
```diff
@@ -280,3 +280,55 @@ bool CScript::HasValidOps() const
     }
     return true;
 }
+
+bool GetScriptOp(CScriptBase::const_iterator& pc, CScriptBase::const_iterator end, opcodetype& opcodeRet, std::vector<unsigned char>* pvchRet)
```
Comment (promag) [inline src/script/script.cpp]: Thread https://github.com/bitcoin/bitcoin/pull/12885/files#r179333509

> make GetScriptOp static.

Agree, force internal linkage.

File: src/script/script.h (hunk 1 above)
Comment (sipa) [reply src/script/script.h]: That defeats the purpose. Longer term is needs to be accessible from the interpreter code.

**Comment by promag** [comment]
Refactor makes sense and looks good to me.

utACK 649b82a.

**Comment by Empact** [comment]
utACK 649b82a

**Comment by ajtowns** [comment]
ACK 649b82a7a58cf1708a4b7134e36e06ea70bdb710

Seems like `CScript::Find()` could be done away with completely as well.

**Comment by Empact** [comment]
@ajtowns yep that was true independent of this PR, opened #12969 to remove

**Comment by sipa** [comment]
Rebased.

**Comment by ajtowns** [comment]
ACK 54a5a21158f740990048b1d43c641630744cf3ee

**Pull request merged**